### PR TITLE
"Start Offline-to-online ingestion" method in Python SDK

### PR DESCRIPTION
**Pull request by pyalex**
**What this PR does / why we need it**:

This PR aim to glue together various backend implementations of job submission and Feast SDK by introducing some cleaner structure with `SparkJob`, `SparkJobParameters` interfaces and separate `launch` modules.
Also `offline_to_online_ingestion` is added to `JobLauncher` interface. Implementations for standalone & dataproc launchers provided as well.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by oavdeev — COMMENTED** [review]
I wonder if there is a way to structure this without the diamond inheritance? After re-reading this a couple times, subjectively i have a pretty hard time following what implements what methods, and what gets called when.

One way I can think of is to maybe split `SparkJob` interface into two:
* `SparkJobParams` that deals with preparing the job to launch (so it provides file path / class name / args) 
* `SparkJob` that deals purely with the job lifecycle after it has been launched (so, id/name/status)

**Comment by pyalex** [comment]
@oavdeev that makes sense

File: sdk/python/feast/pyspark/abc.py [hunk 1]
```diff
@@ -25,6 +35,85 @@ def get_id(self) -> str:
...
+        **kwargs,
+    ):
+        self._feature_tables = feature_tables
+        self._feature_tables_sources = feature_tables_sources
+        self._entity_source = entity_source
+        self._destination = destination
+
+    def get_name(self) -> str:
+        all_feature_tables_names = [ft["name"] for ft in self._feature_tables]
+        return f"HistoryRetrieval-{all_feature_tables_names}"
```
Comment (oavdeev) [inline sdk/python/feast/pyspark/abc.py:98]:
```suggestion
        return f"HistoryRetrieval-{'-'.join(all_feature_tables_names)}"
```

File: sdk/python/feast/pyspark/abc.py [hunk 2]
```diff
@@ -25,6 +35,85 @@ def get_id(self) -> str:
...
+        entity_source: Dict,
+        destination: Dict,
+        **kwargs,
+    ):
+        self._feature_tables = feature_tables
+        self._feature_tables_sources = feature_tables_sources
+        self._entity_source = entity_source
+        self._destination = destination
+
+    def get_name(self) -> str:
```
Comment (oavdeev) [inline sdk/python/feast/pyspark/abc.py:96]: there is a chance that we'll run into limits on name length on different platforms, on EMR it is 256 chars max

File: sdk/python/feast/pyspark/abc.py [hunk 3]
```diff
@@ -53,8 +142,53 @@ def get_output_file_uri(self, timeout_sec=None):
...
 
+class IngestionJobParameters(SparkJobParameters):
+    def __init__(
+        self,
+        feature_table: Dict,
+        source: Dict,
+        start: datetime,
+        end: datetime,
+        jar: str,
+        **kwargs,
```
Comment (oavdeev) [inline sdk/python/feast/pyspark/abc.py:153]: doesn't look like we use kwargs

**Review by khorshuheng — APPROVED** [review]

File: sdk/python/feast/pyspark/launcher.py [hunk 4]
```diff
@@ -123,22 +129,45 @@ def start_historical_feature_retrieval_job(
     job_id: str,
```
Comment (oavdeev) [inline sdk/python/feast/pyspark/launcher.py:129]: i think it would make sense to remove `job_id` here as well

**Review by oavdeev — APPROVED** [review]

File: sdk/python/feast/pyspark/launcher.py (hunk 4 above)
Comment (khorshuheng) [reply sdk/python/feast/pyspark/launcher.py:129]: Noted. Will take note of this in my follow up PR.

Comment (khorshuheng) [reply sdk/python/feast/pyspark/launcher.py]: Removed the job_id

**Pull request merged**